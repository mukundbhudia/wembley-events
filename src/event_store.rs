use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

use chrono::*;
use icalendar::*;

use crate::{SerpapiEvents, WembleyEvent};

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct WembleyEvents {
    pub events: BTreeMap<usize, WembleyEvent>,
}

impl WembleyEvents {
    pub fn new() -> Self {
        WembleyEvents {
            events: BTreeMap::new(),
        }
    }

    pub fn get_events(&self) -> &BTreeMap<usize, WembleyEvent> {
        &self.events
    }

    pub fn build_events_from_html(mut self, html: String) -> WembleyEvents {
        let serp_api_events = serde_json::from_str::<SerpapiEvents>(&html).unwrap_or_default();

        let mut year = serp_api_events.search_metadata.created_at;
        year.truncate(4);

        self.events = serp_api_events
            .events_results
            .into_iter()
            .map(|mut e| {
                let mut swapped_date = e.date.start_date.split_whitespace().collect::<Vec<&str>>();
                swapped_date.push(&year);
                swapped_date.swap(0, 1); // swap [month, day] to [day, month]

                e.date.start_date = swapped_date.join(" ");
                e.into()
            })
            .enumerate()
            .collect::<BTreeMap<usize, WembleyEvent>>();

        Self {
            events: self.events,
        }
    }

    pub fn build_calendar_from_events(self) -> Calendar {
        let mut calendar = Calendar::new();

        self.events.into_iter().for_each(|(_, event)| {
            if let Some(ymd) = event.date_to_ymd() {
                let wembley_event = Event::new()
                    .all_day(Utc.ymd(ymd.year, ymd.month, ymd.day))
                    .summary(&event.title)
                    .description(&event.description)
                    .done();

                calendar.push(wembley_event);
            };
        });

        calendar
    }

    pub fn build_json_from_events(self) -> String {
        serde_json::to_string(&self.events.values().collect::<Vec<_>>()).unwrap()
    }
}

impl Default for WembleyEvents {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use crate::{serpapi_test_output_json_1, serpapi_test_output_json_2};

    use super::*;

    #[test]
    fn build_events_from_html() {
        let body = serpapi_test_output_json_1();
        let wembley_events = WembleyEvents::new().build_events_from_html(body);

        assert_eq!(wembley_events.get_events().len(), 10);
    }

    #[test]
    fn build_calendar_from_events() {
        let body = serpapi_test_output_json_1();
        let wembley_events = WembleyEvents::new().build_events_from_html(body);

        let calendar = wembley_events.build_calendar_from_events();

        assert_eq!(calendar.len(), 10);
    }

    #[test]
    fn check_events_match_calendar() {
        let body = serpapi_test_output_json_1();
        let wembley_events = WembleyEvents::new().build_events_from_html(body);
        let calendar_built_from_events = wembley_events.build_calendar_from_events();

        insta::assert_debug_snapshot!(calendar_built_from_events);
    }

    #[test]
    fn check_events_match_calendar_with_blank_html() {
        let body = serpapi_test_output_json_2();
        let wembley_events = WembleyEvents::new().build_events_from_html(body);
        let calendar_built_from_events = wembley_events.build_calendar_from_events();

        insta::assert_debug_snapshot!(calendar_built_from_events);
    }

    #[test]
    fn check_events_match_calendar_with_blank_string() {
        let body = String::new();
        let wembley_events = WembleyEvents::new().build_events_from_html(body);
        let calendar_built_from_events = wembley_events.build_calendar_from_events();

        insta::assert_debug_snapshot!(calendar_built_from_events);
    }

    #[test]
    fn check_events_match_calendar_json() {
        let body = serpapi_test_output_json_1();
        let wembley_events_as_json = WembleyEvents::new()
            .build_events_from_html(body)
            .build_json_from_events();

        insta::assert_json_snapshot!(wembley_events_as_json);
    }
}
