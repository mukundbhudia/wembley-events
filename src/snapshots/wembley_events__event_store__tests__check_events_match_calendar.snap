---
source: src/event_store.rs
assertion_line: 105
expression: calendar_built_from_events

---
Calendar {
    properties: [],
    components: [
        Event(
            Event {
                inner: InnerComponent {
                    properties: {
                        "DESCRIPTION": Property {
                            key: "DESCRIPTION",
                            val: "Multi-platinum recording artist Harry Styles announces the rescheduled dates for his world tour, in addition to new shows added across the globe. Styles will kick off his colossal 32-city outing...",
                            params: {},
                        },
                        "DTEND": Property {
                            key: "DTEND",
                            val: "20220619",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "DTSTART": Property {
                            key: "DTSTART",
                            val: "20220619",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "SUMMARY": Property {
                            key: "SUMMARY",
                            val: "Harry Styles",
                            params: {},
                        },
                    },
                    multi_properties: [],
                },
            },
        ),
        Event(
            Event {
                inner: InnerComponent {
                    properties: {
                        "DESCRIPTION": Property {
                            key: "DESCRIPTION",
                            val: "Capital's Summertime Ball",
                            params: {},
                        },
                        "DTEND": Property {
                            key: "DTEND",
                            val: "20220612",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "DTSTART": Property {
                            key: "DTSTART",
                            val: "20220612",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "SUMMARY": Property {
                            key: "SUMMARY",
                            val: "David Guetta @ Wembley Stadium",
                            params: {},
                        },
                    },
                    multi_properties: [],
                },
            },
        ),
        Event(
            Event {
                inner: InnerComponent {
                    properties: {
                        "DESCRIPTION": Property {
                            key: "DESCRIPTION",
                            val: "Join BOXPARK Wembley and their resident DJ's for all things soulful",
                            params: {},
                        },
                        "DTEND": Property {
                            key: "DTEND",
                            val: "20220605",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "DTSTART": Property {
                            key: "DTSTART",
                            val: "20220605",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "SUMMARY": Property {
                            key: "SUMMARY",
                            val: "Soulful Sundays",
                            params: {},
                        },
                    },
                    multi_properties: [],
                },
            },
        ),
        Event(
            Event {
                inner: InnerComponent {
                    properties: {
                        "DESCRIPTION": Property {
                            key: "DESCRIPTION",
                            val: "span Ed Sheeran has announced the first leg of his ‘+ - = ÷ x Tour’ (pronounced ‘The Mathematics Tour’) , taking place in stadiums throughout 2022. Kicking off in April next year, the tour will...",
                            params: {},
                        },
                        "DTEND": Property {
                            key: "DTEND",
                            val: "20220629",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "DTSTART": Property {
                            key: "DTSTART",
                            val: "20220629",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "SUMMARY": Property {
                            key: "SUMMARY",
                            val: "Ed Sheeran @ Wembley Stadium",
                            params: {},
                        },
                    },
                    multi_properties: [],
                },
            },
        ),
        Event(
            Event {
                inner: InnerComponent {
                    properties: {
                        "DESCRIPTION": Property {
                            key: "DESCRIPTION",
                            val: "span Ed Sheeran has announced the first leg of his ‘+ - = ÷ x Tour’ (pronounced ‘The Mathematics Tour’) , taking place in stadiums throughout 2022. Kicking off in April next year, the tour will...",
                            params: {},
                        },
                        "DTEND": Property {
                            key: "DTEND",
                            val: "20220625",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "DTSTART": Property {
                            key: "DTSTART",
                            val: "20220625",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "SUMMARY": Property {
                            key: "SUMMARY",
                            val: "Dylan @ Wembley Stadium",
                            params: {},
                        },
                    },
                    multi_properties: [],
                },
            },
        ),
        Event(
            Event {
                inner: InnerComponent {
                    properties: {
                        "DESCRIPTION": Property {
                            key: "DESCRIPTION",
                            val: "Global super pop rock band Coldplay will be hitting the road on tour in 2022 and will be performing live on stage in their Music of the Spheres World Tour ! This particular live concert is...",
                            params: {},
                        },
                        "DTEND": Property {
                            key: "DTEND",
                            val: "20220816",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "DTSTART": Property {
                            key: "DTSTART",
                            val: "20220816",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "SUMMARY": Property {
                            key: "SUMMARY",
                            val: "Coldplay @ Wembley Stadium",
                            params: {},
                        },
                    },
                    multi_properties: [],
                },
            },
        ),
        Event(
            Event {
                inner: InnerComponent {
                    properties: {
                        "DESCRIPTION": Property {
                            key: "DESCRIPTION",
                            val: "Ed Sheeran Seated Ticket & Hotel Experience This Experience Includes: • Reserved Seated Ticket (Block 104 or 141) Tickets seated together • One night hotel stay including complimentary breakfast •...",
                            params: {},
                        },
                        "DTEND": Property {
                            key: "DTEND",
                            val: "20220624",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "DTSTART": Property {
                            key: "DTSTART",
                            val: "20220624",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "SUMMARY": Property {
                            key: "SUMMARY",
                            val: "Ed Sheeran + - = ÷ x Tour",
                            params: {},
                        },
                    },
                    multi_properties: [],
                },
            },
        ),
        Event(
            Event {
                inner: InnerComponent {
                    properties: {
                        "DESCRIPTION": Property {
                            key: "DESCRIPTION",
                            val: "Global super pop rock band Coldplay will be hitting the road on tour in 2022 and will be performing live on stage in their Music of the Spheres World Tour ! This particular live concert is...",
                            params: {},
                        },
                        "DTEND": Property {
                            key: "DTEND",
                            val: "20220820",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "DTSTART": Property {
                            key: "DTSTART",
                            val: "20220820",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "SUMMARY": Property {
                            key: "SUMMARY",
                            val: "Coldplay: Music of the Spheres World Tour",
                            params: {},
                        },
                    },
                    multi_properties: [],
                },
            },
        ),
        Event(
            Event {
                inner: InnerComponent {
                    properties: {
                        "DESCRIPTION": Property {
                            key: "DESCRIPTION",
                            val: "Jubin Nautiyal at OVO Arena Wembley in Wembley Park",
                            params: {},
                        },
                        "DTEND": Property {
                            key: "DTEND",
                            val: "20220626",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "DTSTART": Property {
                            key: "DTSTART",
                            val: "20220626",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "SUMMARY": Property {
                            key: "SUMMARY",
                            val: "Jubin Nautiyal - RESCHEDULED",
                            params: {},
                        },
                    },
                    multi_properties: [],
                },
            },
        ),
        Event(
            Event {
                inner: InnerComponent {
                    properties: {
                        "DESCRIPTION": Property {
                            key: "DESCRIPTION",
                            val: "SJM Concerts 50 CENT FRI, 10 JUN 2022 at 07:30PM BST Doors Open: 06:00PM OnSale: Fri, 28 Jan 2022 at 09:30AM GMT Announcement: Wed, 26 Jan 2022 at 10:00AM GMT Hip Hop heavyweight 50 Cent, has...",
                            params: {},
                        },
                        "DTEND": Property {
                            key: "DTEND",
                            val: "20220610",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "DTSTART": Property {
                            key: "DTSTART",
                            val: "20220610",
                            params: {
                                "VALUE": Parameter {
                                    key: "VALUE",
                                    val: "DATE",
                                },
                            },
                        },
                        "SUMMARY": Property {
                            key: "SUMMARY",
                            val: "50 Cent",
                            params: {},
                        },
                    },
                    multi_properties: [],
                },
            },
        ),
    ],
}
